### PR TITLE
Adapt onCheck event for server side pagination

### PR DESCRIPTION
**Pull request by JSReds**
Change-type: minor
Signed-off-by: Andrea Rosci <andrear@balena.io>

Screenshot 2023-01-05 at 16 44 10
Screenshot 2023-01-05 at 16 44 20

---

##### Contributor checklist

- [ ] I have regenerated jest snapshots for any affected components with `npm run generate-snapshots`

##### Reviewer Guidelines

- When submitting a review, please pick:
  - '_Approve_' if this change would be acceptable in the codebase (even if there are minor or cosmetic tweaks that could be improved).
  - '_Request Changes_' if this change would not be acceptable in our codebase (e.g. bugs, changes that will make development harder in future, security/performance issues, etc).
  - '_Comment_' if you don't feel you have enough information to decide either way (e.g. if you have major questions, or you don't understand the context of the change sufficiently to fully review yourself, but want to make a comment)

---

### CODE REVIEW

File: src/components/Table/TableBase.tsx [hunk 1]
```diff
@@ -377,12 +380,17 @@ export class TableBase<T extends {}> extends React.Component<
...
 		this.setState(newState);
 		if (this.props.onCheck && hasCheckedItemsChanged) {
-			this.props.onCheck(newState.checkedItems);
+			this.props.onCheck(newState.checkedItems, this.state.allChecked);
 		}
 	};
 
+	private toggleCheckedState = () => {
+		const { allChecked } = this.state;
+		return allChecked === 'none' ? 'all' : 'none';
```
Comment (thgreasi) [inline src/components/Table/TableBase.tsx:389]: Would it make sense if this was v ?
```suggestion
		return allChecked === 'all' ? 'none' : 'all';
```
This way:
none -> all
some -> all
all -> none

File: src/components/Table/TableBase.tsx [hunk 2]
```diff
@@ -393,12 +401,14 @@ export class TableBase<T extends {}> extends React.Component<
 					: data.slice();
 		}
 		if (onCheck) {
-			onCheck(checkedItems);
+			onCheck(checkedItems, this.state.allChecked);
```
Comment (thgreasi) [inline src/components/Table/TableBase.tsx:404]: I think that if we are in server side pagination and `this.toggleCheckedState()` returns `all`, then we shouldn't given checkedItems (in which case we also can skip computing them).

Comment (thgreasi) [reply src/components/Table/TableBase.tsx:404]: Let me rephrase this, I think that for server side pagination, giving `checkedItems` & `all` at the same time should only happen when `checkedItems.length === pagination.totalItems`, since that's the only case that `checkedItems` will hold really `all` items.

File: src/components/Table/TableBase.tsx (hunk 1 above)
Comment (JSReds) [reply src/components/Table/TableBase.tsx:389]: that's not what we have today, should we change it to this version ? @thgreasi

Comment (thgreasi) [reply src/components/Table/TableBase.tsx:388]: I hadn't realized this. Ignore this, let's preserve the current behavior.

File: src/components/Table/TableBase.tsx (hunk 2 above)
Comment (JSReds) [reply src/components/Table/TableBase.tsx:408]: yep good point! If you agree I would just pass it as undefined wrapping the computation in a if statement

**Review by thgreasi — APPROVED** [review]

File: src/components/Table/TableBase.tsx [hunk 3]
```diff
@@ -139,11 +139,11 @@ const Base = styled.div<InternalTableBaseProps>`
 	}
 `;
 
-type CheckedTypes = 'none' | 'some' | 'all';
+export type CheckedTypes = 'none' | 'some' | 'all';
```
Comment (thgreasi) [inline src/components/Table/TableBase.tsx:142]: How about CheckedState or something?

File: src/components/Table/TableBase.tsx [hunk 4]
```diff
@@ -592,6 +614,10 @@ export class TableBase<T extends {}> extends React.Component<
 		const checkedRowIdentifiers = this.getCheckedRowIdentifiers();
 		const highlightedRowIdentifiers = this.getHighlightedRowIdentifiers();
 		const disabledRowIdentifiers = this.getDisabledRowIdentifiers();
+		const selectedItems =
```
Comment (thgreasi) [inline src/components/Table/TableBase.tsx:617]:
```suggestion
		const selectedItemCount =
```
